bot/exts/info/patreon.py

```diff
@@ -0,0 +1,162 @@
+import datetime
+import logging
+
+import discord
+from discord.ext import commands, tasks
+
+from bot import constants, utils
+from bot.bot import Bot
+
+log = logging.getLogger(__name__)
+
+
+class Patreon(commands.Cog):
+    """Cog that shows patreon supporters."""
+
+    def __init__(self, bot: Bot) -> None:
+        self.bot = bot
+
+        self.current_monthly_supporters.start()
+
+    @commands.Cog.listener()
+    async def on_member_update(self, before: discord.Member, after: discord.Member) -> None:
+        """Send a message when someone receives a patreon role."""
+        # Ensure the caches are up to date
+        await self.bot.wait_until_guild_available()
+
+        guild = after.guild
+
+        patreon_tier_1_role = guild.get_role(constants.Roles.patreon_tier_1)
+        patreon_tier_2_role = guild.get_role(constants.Roles.patreon_tier_2)
+        patreon_tier_3_role = guild.get_role(constants.Roles.patreon_tier_3)
+
+        current_patreon_tier = 0
+        new_patreon_tier = 0
+
+        # Both of these go from top to bottom to give the user their highest patreon role if they have multiple
+
+        if patreon_tier_3_role in before.roles:
+            current_patreon_tier = 3
+        elif patreon_tier_2_role in before.roles:
+            current_patreon_tier = 2
+        elif patreon_tier_1_role in before.roles:
+            current_patreon_tier = 1
+
+        if patreon_tier_3_role in after.roles:
+            new_patreon_tier = 3
+            colour = patreon_tier_3_role.colour
+        elif patreon_tier_2_role in after.roles:
+            new_patreon_tier = 2
+            colour = patreon_tier_2_role.colour
+        elif patreon_tier_1_role in after.roles:
+            new_patreon_tier = 1
+            colour = patreon_tier_1_role.colour
+
+        if not new_patreon_tier > current_patreon_tier:
+            return
+
+        message = (
+            f":tada: {after.mention} just became a **tier {new_patreon_tier}** patron!\n"
+            "[Support us on Patreon](https://pydis.com/patreon)"
+        )
+
+        sending_channel = utils.channel.get_or_fetch_channel(constants.Channels.meta)
+
+        await sending_channel.send(
+            embed=discord.Embed(
+                description=message,
+                colour=colour
+            )
+        )
+
+    async def send_current_supporters(self, channel: discord.TextChannel) -> None:
+        """Send the current list of patreon supporters, sorted by tier level."""
+        await self.bot.wait_until_guild_available()
+
+        guild = self.bot.get_guild(constants.Guild.id)
+
+        tier_1_or_above_patrons = set(guild.get_role(constants.Roles.patreon_tier_1).members)
+        tier_2_or_above_patrons = set(guild.get_role(constants.Roles.patreon_tier_2).members)
```

Comment:
I like using a set here, since it favours no patreon over another, it effectively randomly sorts them.

Reply:
Most certainly not the intended reason for this set usage :-)